neo4j/driver_with_context_test.go

```diff
@@ -0,0 +1,507 @@
+				}},
+			expectedSessionConfig: defaultSessionConfig,
+			expectedErr:           fmt.Errorf("dude, where are my keys"),
+		},
+		{
+			description: "returns error when read result records cannot be collected",
+			configurers: []ExecuteQueryConfigurationOption{WithReadersRouting()},
+			createSession: &fakeSession{
+				executeReadTransactionResult: &fakeResult{
+					keys:       keys,
+					collectErr: fmt.Errorf("one does not simply collect"),
+				}},
+			expectedSessionConfig: defaultSessionConfig,
+			expectedErr:           fmt.Errorf("one does not simply collect"),
+		},
+		{
+			description: "returns error when read result summary cannot be retrieved",
+			configurers: []ExecuteQueryConfigurationOption{WithReadersRouting()},
+			createSession: &fakeSession{
+				executeReadTransactionResult: &fakeResult{
+					keys:       keys,
+					collect:    records,
+					summaryErr: fmt.Errorf("in summary: nope"),
+				}},
+			expectedSessionConfig: defaultSessionConfig,
+			expectedErr:           fmt.Errorf("in summary: nope"),
+		},
+		{
+			description: "returns error when read result summary cannot be retrieved",
+			configurers: []ExecuteQueryConfigurationOption{WithReadersRouting()},
+			createSession: &fakeSession{
+				executeReadTransactionResult: &fakeResult{
+					keys:       keys,
+					collect:    records,
+					summaryErr: fmt.Errorf("in summary: nope"),
+				}},
+			expectedSessionConfig: defaultSessionConfig,
+			expectedErr:           fmt.Errorf("in summary: nope"),
+		},
+		{
+			description: "returns error when read execution fails",
+			configurers: []ExecuteQueryConfigurationOption{WithReadersRouting()},
+			createSession: &fakeSession{
+				executeReadErr: fmt.Errorf("oopsie"),
+			},
+			expectedSessionConfig: defaultSessionConfig,
+			expectedErr:           fmt.Errorf("oopsie"),
+		},
+	}
+
+	for _, testCase := range testCases {
+		outer.Run(testCase.description, func(t *testing.T) {
+			driver := &driverWithContext{
+				newSession: func(_ context.Context, config SessionConfig) SessionWithContext {
+					AssertDeepEquals(t, testCase.expectedSessionConfig, config)
+					return testCase.createSession
+				},
+				defaultManagedSessionBookmarkManager: defaultBookmarkManager,
+			}
+
+			eagerResult, err := driver.ExecuteQuery(ctx, "RETURN 42", nil, testCase.configurers...)
+
+			AssertDeepEquals(t, testCase.expectedErr, err)
+			AssertDeepEquals(t, testCase.expectedResult, eagerResult)
+		})
+	}
+
+	outer.Run("default bookmark manager is thread-safe", func(t *testing.T) {
+		driver := &driverWithContext{
+			newSession: func(_ context.Context, config SessionConfig) SessionWithContext {
+				return &fakeSession{
+					executeWriteErr: fmt.Errorf("oopsie, write failed"),
+				}
+			},
+		}
+
+		var wait sync.WaitGroup
+		var bookmarkManagerAddresses sync.Map
+		goroutineCount := 50
+		wait.Add(goroutineCount)
+		for i := 0; i < goroutineCount; i++ {
+			go func(i int) {
+				callExecuteQueryOrBookmarkManagerGetter(driver, i)
```

Comment:
Might be worth to run this test twice. Once with `Driver.GetDefaultManagedBookmarkManager` being called first and once with `driver.ExecuteQuery` first.